### DISCUSSION TITLE
Osu!Lazer running horribly compared to its stable counterpart.

### DISCUSSION THREAD

**Discussion by freqawi** [ppy/osu#28873] [Q&A] [answered]
### Type

Performance

### Bug description

I am writing on behalf of a friend who is not comfortable with github, 

I have been trying to troubleshoot with them on the issues, however nothing I have tried, (from web results and my own knowledge base), has worked. I had the same issue they had a long time ago, but I thought it might have been fixed by now.

They are running a laptop, ROG Zephyrus M15 with GTX 1660 and i7-10750H, Latest Nvidia Driver (556.12), Windows 22H2 19045.4651. 

The Issue: Osu Lazer runs at very low framerate compared to stable, on their laptop display, Osu lazer is shown to use GPU 0 (Intel UHD) when running, but osu stable uses GPU 1 (Nvidia 1660) and runs just fine (With Compatibilty mode enabled)

I have been unsuccessful in getting the process to run with the correct GPU.

I've tried;

1. Setting Osu!lazer to highest performance in Windows Graphics.
2. Settings Osu!Lazer to highest performance in Nvidia Control panel
3. Reinstalling Driver
4. Tried all renderers the game has to offer, no large performance uplift.
5. Made sure both WIndows and Armory Crate are running in their highest presets.

It seems like when you attempt to add both Stable and Lazer to nvidia control panel they get merged(?) and do not show as separate executable's.

### Screenshots or videos

![image](https://github.com/user-attachments/assets/6b31a777-d16e-4d6f-b2aa-3b20c12e2248)

### Version

2024-625.2-lazer

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/16233749/compressed-logs.zip)

**Comment by peppy** [comment]
Can you try running lazer from the versioned folder rather than the main `osu!.exe` and see if that helps? It will be in a folder next to the main linked executable.

Also, please switch off the experimental renderer. This is going to net you worse performance guaranteed.

**Comment by freqawi** [comment]
Apologies for the late reply, 

Running the executable in the APP directory didn't fix the issue and task manager still shows using GPU0 and the latency in game is still high.

"fps is roughly the same, the ms thing fluctuates around 1.2-2ish ms. i don't think it made a difference."
"little to no difference."

While also using the non experimental DirectX11.

I've had him add that specific EXE to Nvidia Control Panel for high performance and within windows GPU performance settings like I've stated before, When adding it to nvidia, it continued the behaivour of showing a 'merged' settings with lazer and stable. And when adding to Windows GPU settings, it stayed the same.

**Reply by peppy** [reply]
1.2-2 ms is already above any requirements. Can you better describe what the actual perceived issue is?

**Reply by freqawi** [reply]
Ofcourse, the issue is that it is using the wrong gpu and can achieve better stable performance with the 1660 as currently it is fluctuating within that range compared to being consistently at 1ms or close to.

**Reply by peppy** [reply]
We definitely want to get it running on the correct GPU, but would also hope it runs okay without doing so. But let's leave that for now.

@smoogipoo thought the squirrel executable may be tricking windows into using the non-dedicated GPU but we've ruled that out now. Will need to figure why this is happening.

**Reply by freqawi** [reply]
I don't know why but my brain wants to figure out why the executables are showing merged in nvidia. Unless that is a known non issue as when I was on my last system adding executables to nvidia would always show their path in Windows as their title, but instead they both just show as "Osu!"

**Reply by peppy** [reply]
nvidia from my recollection is only based on executable name. you can try renaming the lazer executable temporarily and test this.

**Reply by freqawi** [reply]
I did already try this, on my system, both stable and lazer refused to launch after renaming their executables, I didn't try the 'app' directory one as this was in the initial troubleshooting

**Reply by freqawi** [reply]
So I asked them to rename the executable and add it to nvidia and windows GPU

"fps is better. now averaging 280-300fps. latency shows around 2.2-3.5ms, but the draw latency itself is 3ms-4ms. improvement, maybe able to improve it more with some tweaking to the settings"

"something that's weird tho is that when I check task manager, the program for lazer sometimes displays as using the igpu and sometimes it displays the dgpu"

**Reply by smoogipoo** [reply]
I think we really have to wait until the migration to Velopack to really understand more about what's going on here :/

Will try to repro on one of the PCs we have.

**Reply by freqawi** [reply]
I definitely know what this means and am completely on board.

**Comment by smoogipoo** [comment]
Try checking your power profiles.
- Do you have power plugged into your laptop?
- Are you using the power saving profile? Or balanced? Maybe try using a higher performance one.

**Reply by freqawi** [reply]
I did this in troubleshooting. I just forgot to put it in (I was tired)

They were using highest performance in Windows, armory crate and was plugged in.

Edit: nvm I actually did include this troubleshooting in the post, maybe you had missed it.

**Reply by smoogipoo** [reply]
Unsure atm unfortunately. Am not able to repro on my Optimus machine - it selects the dGPU by default but I can set the versioned exe (`AppData/osulazer/<version>/osu.exe`) to use the iGPU in the Windows settings and it'll work as expected.

**Reply by freqawi** [reply]
It is very odd tbh, given stable uses the dGPU always, I thought the behavior of swapping from showing dGPU and iGPU in task manager might be because of the way optimus handles the frames, being rendered on dGPU then processed through iGPU to be sent to the screen.

Do you get this same occurrence?

**Reply by smoogipoo** [reply]
Oh, I've totally misunderstood this entire thread, sorry.

Now that I take a look at your logs again, osu! is indeed running with the dGPU. The iGPU being active is intentional as it's the processor displaying the frame to the screen (so you'll see high Copy% on the dGPU and high 3D% on the iGPU). This is the same as osu!stable.

What task manager's displaying is likely the processor with highest 3D%, which is different between lazer and stable likely because they just perform differently.

**Reply by freqawi** [reply]
I think the issue still "remains" or its likely the underperformance of laptop hardware.
This is what I asked them about the difference of the renamed executable.
"before was around 200fps with 3-4ms or 3-5ms draw latency
after was around 280-300fps with 3-4ms draw"

I thought draw latency would be quite a bit lower with than increased FPS rather than being roughly the same.

**Reply by rXelelo** [reply]
on windows i have problem with fps but after i go in windowed i have boost +100

**Reply by smoogipoo** [reply, marked as answer]
There's nothing actionable here. Performance is still a work in progress.

**Reply by freqawi** [reply]
alright thank you, 

on another note irrelevant to this, I want to make a dump of features or changes I would like for lazer, is this something i can do and if so where?

**Reply by smoogipoo** [reply]
Please check existing open issues & discussions, and open a new discussion: https://github.com/ppy/osu/discussions/new/choose
